core/src/main/scala/kafka/server/KafkaApis.scala

```diff
@@ -1520,29 +1522,66 @@ class KafkaApis(val requestChannel: RequestChannel,
   def handleLeaveGroupRequest(request: RequestChannel.Request) {
     val leaveGroupRequest = request.body[LeaveGroupRequest]
 
-    // the callback for sending a leave-group response
-    def sendResponseCallback(error: Errors) {
-      def createResponse(requestThrottleMs: Int): AbstractResponse = {
-        val response = new LeaveGroupResponse(new LeaveGroupResponseData()
-          .setThrottleTimeMs(requestThrottleMs)
-          .setErrorCode(error.code()))
-        trace("Sending leave group response %s for correlation id %d to client %s."
-          .format(response, request.header.correlationId, request.header.clientId))
-        response
-      }
-      sendResponseMaybeThrottle(request, createResponse)
+    // Before version 3, leave group request is still in single mode
+    val members = if (leaveGroupRequest.version <= 2) {
+      List(new MemberIdentity()
+        .setMemberId(leaveGroupRequest.data.memberId))
+    } else {
+      leaveGroupRequest.data.members.asScala.toList
     }
 
-    if (!authorize(request.session, Read, Resource(Group, leaveGroupRequest.data().groupId(), LITERAL))) {
-      sendResponseMaybeThrottle(request, requestThrottleMs =>
+    if (!authorize(request.session, Read, Resource(Group, leaveGroupRequest.data.groupId, LITERAL))) {
+      sendResponseMaybeThrottle(request, requestThrottleMs => {
         new LeaveGroupResponse(new LeaveGroupResponseData()
           .setThrottleTimeMs(requestThrottleMs)
-          .setErrorCode(Errors.GROUP_AUTHORIZATION_FAILED.code())))
+          .setErrorCode(Errors.GROUP_AUTHORIZATION_FAILED.code)
+          .setMembers(
+            members.map (
+              member => new MemberResponse()
+                .setMemberId(member.memberId)
+                .setGroupInstanceId(member.groupInstanceId)
+                .setErrorCode(Errors.GROUP_AUTHORIZATION_FAILED.code)).asJava
+          )
+        )
+      })
     } else {
-      // let the coordinator to handle leave-group
+      def sendResponseCallback(leaveGroupResults: List[LeaveGroupResult]) {
+        val memberResponses = leaveGroupResults.map(
+          leaveGroupResult =>
+            new MemberResponse()
+              .setErrorCode(leaveGroupResult.error.code)
+              .setMemberId(leaveGroupResult.memberId)
+              .setGroupInstanceId(leaveGroupResult.groupInstanceId.orNull)
+        )
+        def createResponse(requestThrottleMs: Int): AbstractResponse = {
+          val batchLeaveGroupResponse = if (leaveGroupRequest.version <= 2) {
+            if (leaveGroupResults.size != 1) {
+              throw new IllegalStateException(s"Singleton leave group request shouldn't have more than 1 " +
+                s"response, while actually get ${leaveGroupResults.size}")
```

Comment:
nit: last part is not grammatical

Reply:
I change 1 -> one, let me know your thoughts.